BuddySave/CloudManager.cs

```diff
@@ -2,14 +2,41 @@ namespace BuddySave;
 
 public class CloudManager : ICloudManager
 {
+    private readonly IBackupDirectoryProvider _backupDirectoryProvider;
+
+    public CloudManager(IBackupDirectoryProvider backupDirectoryProvider)
+    {
+        _backupDirectoryProvider = backupDirectoryProvider;
+    }
+    
     public void UploadSave(GameSave gameSave)
     {
-        CopyOverSaves(gameSave.LocalPath, gameSave.CloudPath);
+        ValidateSource(gameSave.LocalPath);
+        BackupFiles(gameSave.CloudPath, gameSave.Name, SaveType.Cloud);
+
+        try
+        {
+            CopyOverSaves(gameSave.LocalPath, gameSave.CloudPath);
+        }
+        catch (Exception)
+        {
+            RestoreBackup(gameSave.CloudPath, gameSave.Name, SaveType.Cloud);
+        }
     }
 
     public void DownloadSave(GameSave gameSave)
     {
-        CopyOverSaves(gameSave.CloudPath, gameSave.LocalPath);
+        ValidateSource(gameSave.CloudPath);
```

Comment:
The whole flow is the same. Could be wrapped up into one method:
void SecureCopy(source, destination, destinationType)
void ProtectedCopy(source, destination, destinationType)
void ManagedCopy(source, destination, destinationType)
?

Reply:
I think I might do some simplification here, since copy/pasting so many tests is getting ridiculous 😆